### DISCUSSION TITLE
Rate Limiting

### DISCUSSION THREAD

**Discussion by jqnatividad** [ckan/ckan#6581] [Ideas]
In the traditional "open data" use case, CKAN operators publish data as a public good.

However, their bandwidth budget can often be exhausted by a few crawlers and bad actors that repeatedly download data without taking this bandwidth limitation into account.

What rate-limiting mechanisms have CKAN operators deployed to handle this?

Can we bake in simple rate-limiting in CKAN, perhaps, by taking advantage of stuff like [Flask-Limiter](https://flask-limiter.readthedocs.io/en/2.x/index.html) and [HTTP RateLimit header fields](https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html) and getting popular CKAN clients like [ckanapi](https://github.com/ckan/ckanapi) to support rate limits?

**Comment by frafra** [comment]
I would suggest offloading such functionality to the web server, like NGINX or Apache, or any kind of reverse proxy used to expose your service to the wide public. Extending CKAN default configuration files could be beneficial.

**Reply by jqnatividad** [reply]
You're right. For simple site-wide rate-limits, its best to offload to the web server and we should add some information about that in docs for commonly used servers.

However, for more granular rate-limiting, using Flask-Limiter still merits investigation:
* Ability to optionally specify rate-limit per API Token/User/Organization in priority order.
* Ability to have different rate-limiting strategies based on schedule, ip address (blacklist/whitelist/country of origin, etc.)
* We can also leverage rate-limiting for enhanced security
    * for highly-sensitive, private packages - e.g. only allowing internal IP addresses/certain subnets for private packages. 
    * You can even possibly control access based on users/collaborators/organizations ala AWS Security Policy
